### PR TITLE
REP-6607 Multi tenant recipe

### PR DESCRIPTION
**Pull request by adrianjgeorge**

### CODE REVIEW

File: repose-aggregator/src/docs/asciidoc/recipes/multi-tenant.adoc [hunk 1]
```diff
@@ -0,0 +1,99 @@
+= Implementing Multi-tenant Authorization
+
+This recipe shows you how to configure your filters to support rest resources that contain multiple tenants in the uri
+all of which need authorization applied to them.
+
+For the remainder of this recipe we are going to assume that you are trying to secure a resource this resource: `/a_tenant/do/something/on/another_tenant`.
```
Comment (wdschei) [inline repose-aggregator/src/docs/asciidoc/recipes/multi-tenant.adoc]: `a resource` needs removed.

File: repose-aggregator/src/docs/asciidoc/recipes/multi-tenant.adoc [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+== Derive Tenants
+
+The first thing to do is to use the <<../filters/url-extractor-to-header.adoc#,URL Extractor to Header filter>> to pull out both tenants we care about into a common header.
```
Comment (wdschei) [inline repose-aggregator/src/docs/asciidoc/recipes/multi-tenant.adoc]: If we are going to reference the `URL Extractor to Header filter` in this recipe, then it should probably be fleshed out too.

File: repose-aggregator/src/docs/asciidoc/recipes/multi-tenant.adoc [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        </validate-tenant>
+    </tenant-handling>
+    <require-service-endpoint <!--2-->
+        public-url="https://service.example.com"
+        region="ORD"
+        name="OpenStackCompute"
+        type="compute"/>
+</keystone-v2-authorization>
+----
+<1> The header that contains the tenants to verify. This has to match the name provided in the Url Extractor Header filter.+
```
Comment (wdschei) [inline repose-aggregator/src/docs/asciidoc/recipes/multi-tenant.adoc]: There needs to be a space between the `.` and the `+` for the formatting to work properly.

File: repose-aggregator/src/docs/asciidoc/recipes/multi-tenant.adoc [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    Every value found in this header has to be present among those returned by keystone.
+<2> The endpoint that you want to verify that the user has, note that this only works if you configure authentication to send the catalog.
+
+== Configure Role Based Access Control
+
+To get per tenant based access control you need to configure the <<../filters/api-validator.adoc#,API Validator filter>> nad the accompanying wadl.
```
Comment (wdschei) [inline repose-aggregator/src/docs/asciidoc/recipes/multi-tenant.adoc]: `nad` => `and`

Comment (wdschei) [reply repose-aggregator/src/docs/asciidoc/recipes/multi-tenant.adoc]: This should also probably link to `<<./role-based-access-control.adoc#3-utilize-tenanted-rax-roles,Role Based Access Control with multiple teanant>>` so the nuance about casing is brought to their attention.

File: repose-aggregator/src/docs/asciidoc/filters/url-extractor-to-header.adoc [hunk 5]
```diff
@@ -1,52 +1,63 @@
...
+    <extraction header="X-Server-Id" url-regex=".*/servers/([^/]+).*"/> <!--4-->
+
+    <extraction header="X-Filter-Param" url-regex=".*\?.*filter=([^&amp;]+).*" default="none"/> <!--5-->
+</url-extractor-to-header>
 ----
-<1> Specifies ???.
+<1> The name of the header to populate.
+<2> The regex to evaluate against the request url. +
+    The capture group is required and is the value that will be placed in the header. +
+    *Note*: Only the first capture group will be populated into the header.
```
Comment (wdschei) [inline repose-aggregator/src/docs/asciidoc/filters/url-extractor-to-header.adoc]: To stay with the admonitions scheme in use elsewhere, this line should be changed to:
```
+
[NOTE]
====
Only the first capture group will be populated into the header.
====
```

Comment (adrianjgeorge) [reply repose-aggregator/src/docs/asciidoc/filters/url-extractor-to-header.adoc]: note's don't work in callouts i thought?

Comment (wdschei) [reply repose-aggregator/src/docs/asciidoc/filters/url-extractor-to-header.adoc]: They do with the extra line that only contains a `+` between the callout and the admonition.

Comment (adrianjgeorge) [reply repose-aggregator/src/docs/asciidoc/filters/url-extractor-to-header.adoc]: Ugh and it has to be left aligned. le sigh

**Pull request merged**